internal/fs/vss_windows.go

```diff
@@ -0,0 +1,980 @@
+		timeoutInMillis)
+	if err != nil {
+		// After calling PrepareForBackup one needs to call AbortBackup() before releasing the VSS
+		// instance for proper cleanup.
+		// It is not neccessary to call BackupComplete before releasing the VSS instance afterwards.
+		iVssBackupComponents.AbortBackup()
+		iVssBackupComponents.Release()
+		return VssSnapshot{}, err
+	}
+
+	err = callAsyncFunctionAndWait(iVssBackupComponents.DoSnapshotSet, "DoSnapshotSet",
+		timeoutInMillis)
+	if err != nil {
+		iVssBackupComponents.AbortBackup()
+		iVssBackupComponents.Release()
+		return VssSnapshot{}, err
+	}
+
+	var snapshotProperties VssSnapshotProperties
+	err = iVssBackupComponents.GetSnapshotProperties(snapshotSetID, &snapshotProperties)
+	if err != nil {
+		iVssBackupComponents.AbortBackup()
+		iVssBackupComponents.Release()
+		return VssSnapshot{}, err
+	}
+
+	for mountPoint, info := range mountPointInfo {
+		err := iVssBackupComponents.GetSnapshotProperties(info.snapshotSetID,
+			&info.snapshotProperties)
+		if err != nil {
+			msgError(mountPoint, errors.Errorf(
+				"VSS error: GetSnapshotProperties() for mount point %s returned error: ",
+				mountPoint, err))
+			info.isSnapshotted = false
+		} else {
+			info.snapshotDeviceObject = info.snapshotProperties.GetSnapshotDeviceObject()
+		}
+
+		mountPointInfo[mountPoint] = info
+	}
+
+	return VssSnapshot{iVssBackupComponents, snapshotSetID, snapshotProperties,
+		snapshotProperties.GetSnapshotDeviceObject(), mountPointInfo, timeoutInMillis}, nil
+}
+
+// Delete deletes the created snapshot.
+func (p *VssSnapshot) Delete() error {
+	var err error
+	if err = vssFreeSnapshotProperties(&p.snapshotProperties); err != nil {
+		return err
+	}
+
+	for _, mountPoint := range p.mountPointInfo {
+		if mountPoint.isSnapshotted {
+			if err = vssFreeSnapshotProperties(&mountPoint.snapshotProperties); err != nil {
+				return err
+			}
+		}
+	}
+
+	if p.iVssBackupComponents != nil {
+		defer p.iVssBackupComponents.Release()
+
+		err = callAsyncFunctionAndWait(p.iVssBackupComponents.BackupComplete, "BackupComplete",
+			p.timeoutInMillis)
+
+		if _, _, err := p.iVssBackupComponents.DeleteSnapshots(p.snapshotID); err != nil {
+			err = newVssTextError(fmt.Sprintf("Failed to delete snapshot:  %s", err.Error()))
+			p.iVssBackupComponents.AbortBackup()
+		}
+	}
+
+	return err
+}
+
+// asyncCallFunc is the callback type for callAsyncFunctionAndWait.
+type asyncCallFunc func() (*IVSSAsync, error)
+
+// callAsyncFunctionAndWait calls an async functions and waits for it to either
+// finish or timeout.
+func callAsyncFunctionAndWait(function asyncCallFunc, name string, timeoutInMillis uint32) error {
+	if iVssAsync, err := function(); err != nil {
+		return err
+	} else {
+		if iVssAsync == nil {
+			return newVssTextError(fmt.Sprintf("%s() returned nil", name))
+		}
+
+		err := iVssAsync.WaitUntilAsyncFinished(timeoutInMillis)
+		iVssAsync.Release()
+		return err
+	}
+
+	return nil
```

Comment:
I think this `return` is unreachable.

Reply:
Fixed in https://github.com/restic/restic/pull/2274/commits/b3a09695590e7c614d50c9aca1b41c8fdda83cc7